pyartcd/pyartcd/pipelines/promote.py

```diff
@@ -0,0 +1,628 @@
+        if not isinstance(advisory_info, dict):
+            raise ValueError(f"Got invalid advisory info for advisory {advisory}: {advisory_info}.")
+        return advisory_info
+
+    @staticmethod
+    def get_live_id(advisory_info: Dict):
+        # Extract live ID from advisory info
+        # Examples:
+        # - advisory with a live ID
+        #     "errata_id": 2681,
+        #     "fulladvisory": "RHBA-2019:2681-02",
+        #     "id": 46049,
+        #     "old_advisory": "RHBA-2019:46049-02",
+        # - advisory without a live ID
+        #     "errata_id": 46143,
+        #     "fulladvisory": "RHBA-2019:46143-01",
+        #     "id": 46143,
+        #     "old_advisory": null,
+        if advisory_info["errata_id"] == advisory_info["id"]:
+            # advisory doesn't have a live ID
+            return None
+        live_id = advisory_info["fulladvisory"].rsplit("-", 1)[0]  # RHBA-2019:2681-02 => RHBA-2019:2681
+        return live_id
+
+    def verify_image_advisory(self, advisory_info: Dict):
+        issues = []
+        live_id = self.get_live_id(advisory_info)
+        if not live_id:
+            issues.append(f"Advisory {advisory_info['id']} doesn't have a live ID.")
+        if advisory_info["status"] not in {"QE", "REL_PREP", "PUSH_READY", "IN_PUSH", "SHIPPED_LIVE"}:
+            issues.append(f"Advisory {advisory_info['id']} cannot be in {advisory_info['status']} state.")
+        if issues:
+            raise VerificationError(f"Advisory {advisory_info['id']} has the following issues:\n" + '\n'.join(issues))
+
+    async def verify_attached_bugs(self, advisories: Iterable[int]):
+        advisories = list(advisories)
+        if not advisories:
+            self._logger.warning("No advisories to verify.")
+            return
+        cmd = [
+            "elliott",
+            f"--group={self.group}",
+            "verify-attached-bugs",
+            "--"
+        ]
+        for advisory in advisories:
+            cmd.append(f"{advisory}")
+        async with self._elliott_lock:
+            await exectools.cmd_assert_async(cmd, env=self._elliott_env_vars, stdout=sys.stderr)
+
+    async def promote_all_arches(self, release_name: str, arches: List[str], previous_list: List[str], metadata: Optional[Dict], reference_releases: Dict[str, str], tag_stable: bool):
+        futures = []
+        for arch in arches:
+            futures.append(self.promote_arch(release_name, arch, previous_list, metadata, reference_releases.get(arch), tag_stable))
+        try:
+            release_infos = await asyncio.gather(*futures)
+        except ChildProcessError as err:
+            self._logger.error("Error promoting release images: %s\n%s", str(err), traceback.format_exc())
+            raise
+        return release_infos
+
+    async def promote_arch(self, release_name: str, arch: str, previous_list: List[str], metadata: Optional[Dict], reference_release: Optional[str], tag_stable: bool):
```

Comment:
I'd love to see tests for this method. Which is not a merge blocker, but let's file a card.

Reply:
😆 I intended to add unit tests for all methods in the pipeline class, but I haven't done that yet. Let me pick the most important ones and add them first.